### PR TITLE
Adding support for text elements to be used as test Data for mapping test flow

### PR DESCRIPTION
**Pull request by gs-gunjan**
## Summary

Adding support for text elements to be used as test Data for mapping test flow.
Refactoring the mapping test flow to use the new end-point
allowing text elements to be passed as request body for resolution

### CODE REVIEW

**Comment by MauricioUyaguari** [comment]
Can we get screenshots or video snippets if there are UI changes please. See other MRs for examples. Thanks

File: .changeset/proud-ducks-admire.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+'@finos/legend-extension-dsl-text': patch
+---
+
+Make text Packageable elements part of the pruned graphData sent to requests
```
Comment (MauricioUyaguari) [inline .changeset/proud-ducks-admire.md:5]:
````suggestion
Add `text` packagable elements to execution input getter to include in `engine` execution requests. 
````

File: .changeset/silly-bananas-dream.md [hunk 2]
```diff
@@ -0,0 +1,6 @@
+---
+'@finos/legend-studio': major
+'@finos/legend-manual-tests': patch
```
Comment (MauricioUyaguari) [inline .changeset/silly-bananas-dream.md:3]: @finos/legend-manual-tests should be its own changeset without any comments.

File: .changeset/silly-bananas-dream.md [hunk 3]
```diff
@@ -0,0 +1,6 @@
+---
+'@finos/legend-studio': major
```
Comment (MauricioUyaguari) [inline .changeset/silly-bananas-dream.md:2]: Should be a patch

File: packages/legend-studio/src/stores/editor-state/element-editor-state/mapping/MappingTestState.ts [hunk 4]
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable @finos/legend-studio/no-cross-workspace-source-usage */
```
Comment (MauricioUyaguari) [inline packages/legend-studio/src/stores/editor-state/element-editor-state/mapping/MappingTestState.ts:1]: Let's remove this and fix the import by exporting the required type from its relevant package

File: packages/legend-studio/src/stores/editor-state/element-editor-state/mapping/MappingTestState.ts [hunk 5]
```diff
@@ -84,6 +81,7 @@ import {
   PureClientVersion,
 } from '@finos/legend-graph';
 import { LambdaEditorState, TAB_SIZE } from '@finos/legend-application';
+import type { MappingTestResult } from '@finos/legend-graph/src/graphManager/action/execution/ExecutionResult';
```
Comment (MauricioUyaguari) [inline packages/legend-studio/src/stores/editor-state/element-editor-state/mapping/MappingTestState.ts:84]: Need to ensure legend-graph exports `MappingTestResult`

File: packages/legend-studio/src/stores/editor-state/element-editor-state/mapping/MappingTestState.ts [hunk 6]
```diff
@@ -624,35 +618,22 @@ export class MappingTestState {
...
-        // TODO: this logic should probably be better handled in by engine mapping test runner
-        assertionMatched =
-          hashObject(extractExecutionResultValues(result)) ===
-          hashObject(losslessParse(this.assertionState.expectedResult));
-      } else {
-        throw new UnsupportedOperationError();
-      }
       this.setResult(
-        assertionMatched ? TEST_RESULT.PASSED : TEST_RESULT.FAILED,
+        result.result === 'SUCCESS' ? TEST_RESULT.PASSED : TEST_RESULT.FAILED,
```
Comment (MauricioUyaguari) [inline packages/legend-studio/src/stores/editor-state/element-editor-state/mapping/MappingTestState.ts:636]: Let's use the correct enum and not the string `SUCCESS`

File: packages/legend-graph/src/graphManager/action/execution/ExecutionResult.ts [hunk 7]
```diff
@@ -31,6 +31,11 @@ export class ResultBuilder {
   }
 }
 
+export class MappingTestResult {
+  actual!: string;
+  result!: string;
```
Comment (MauricioUyaguari) [inline packages/legend-graph/src/graphManager/action/execution/ExecutionResult.ts:36]: Shouldn't this be an Enum?

File: packages/legend-graph/src/models/protocols/pure/v1/engine/execution/V1_ExecuteMappingTestInput.ts [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+import { V1_pureModelContextDataPropSchema } from '../../transformation/pureProtocol/V1_PureProtocolSerialization';
+import type { V1_RawExecutionContext } from '../../model/rawValueSpecification/V1_RawExecutionContext';
+import { SerializationFactory, usingModelSchema } from '@finos/legend-shared';
+import { createModelSchema, optional, primitive } from 'serializr';
+import { V1_rawBaseExecutionContextModelSchema } from '../../transformation/pureProtocol/serializationHelpers/V1_RawValueSpecificationSerializationHelper';
+
+export class V1_ExecuteMappingTestInput {
+  clientVersion!: string;
+  mapping!: string;
+  testId!: string;
```
Comment (MauricioUyaguari) [inline packages/legend-graph/src/models/protocols/pure/v1/engine/execution/V1_ExecuteMappingTestInput.ts:27]: I thought we said `testID` would be optional. If optional then run all the tests?

Comment (gs-gunjan) [reply packages/legend-graph/src/models/protocols/pure/v1/engine/execution/V1_ExecuteMappingTestInput.ts:27]: we said we want testId to be a list , but if we want to run all then we send nothing in the list, but what we can do instead is just send 'all' in the testId and we can provide mechanism in the backend to run all tests.
Currently didnt implement that, there is another thing of what response to send in this case and how studio will display it

**Comment by gs-gunjan** [comment]
also wanted to delete the runtime code in MappingTestState.ts, but it seems to be used for generating plan for a mapping test

File: packages/legend-studio/src/stores/editor-state/element-editor-state/mapping/MappingTestState.ts (hunk 4 above)
Comment (gs-gunjan) [reply packages/legend-studio/src/stores/editor-state/element-editor-state/mapping/MappingTestState.ts]: done

File: packages/legend-studio/src/stores/editor-state/element-editor-state/mapping/MappingTestState.ts (hunk 5 above)
Comment (gs-gunjan) [reply packages/legend-studio/src/stores/editor-state/element-editor-state/mapping/MappingTestState.ts]: yeh, added it to the export list

File: packages/legend-studio/src/stores/editor-state/element-editor-state/mapping/MappingTestState.ts (hunk 6 above)
Comment (gs-gunjan) [reply packages/legend-studio/src/stores/editor-state/element-editor-state/mapping/MappingTestState.ts:635]: We get this using legend-engine, so will make the change there

File: packages/legend-graph/src/models/protocols/pure/v1/engine/execution/V1_ExecuteMappingTestInput.ts (hunk 8 above)
Comment (MauricioUyaguari) [reply packages/legend-graph/src/models/protocols/pure/v1/engine/execution/V1_ExecuteMappingTestInput.ts:27]: I think we should do the list. If the lsit is empty or null run the full test suit. We should return the results as a list per each test run.

File: packages/legend-graph/src/models/protocols/pure/v1/engine/V1_EngineServerClient.ts [hunk 9]
```diff
@@ -286,6 +290,21 @@ export class V1_EngineServerClient extends AbstractServerClient {
       { skipProcessing: Boolean(returnResultAsText) },
     );
 
+  doMappingTest = (
+    input: PlainObject<V1_ExecuteMappingTestInput>,
+    returnResultAsText?: boolean,
+  ): Promise<PlainObject<V1_MappingTestResult> | Response> =>
```
Comment (MauricioUyaguari) [inline packages/legend-graph/src/models/protocols/pure/v1/engine/V1_EngineServerClient.ts:296]: Granted have not reviewed legend engine changes yet (will do asap)
But the API should return a lsit of `MappingTestResult` since it should be able to run, all, some or one tests. 
Therefore the return type should be `Promise<PlainObject<V1_MappingTestResult>[]>`

File: packages/legend-graph/src/models/protocols/pure/v1/model/packageableElements/store/relational/mapping/V1_RelationalInputData.ts [hunk 10]
```diff
@@ -18,18 +18,19 @@ import { hashArray } from '@finos/legend-shared';
 import type { Hashable } from '@finos/legend-shared';
 import { CORE_HASH_STRUCTURE } from '../../../../../../../../../MetaModelConst';
 import { V1_InputData } from '../../../../../model/packageableElements/mapping/V1_InputData';
+import type { V1_TestDataSource } from '../../modelToModel/mapping/V1_TestDataSource';
 
 export class V1_RelationalInputData extends V1_InputData implements Hashable {
   database!: string;
-  data!: string;
```
Comment (MauricioUyaguari) [inline packages/legend-graph/src/models/protocols/pure/v1/model/packageableElements/store/relational/mapping/V1_RelationalInputData.ts:24]: We should still be able to read this property field, i.e be backward compatible. 
So we should read this property but during serialization we write this as 
`V1_StringTestDataSource` if the data property has been filled. 
The key is in the builder if the data property is present we build the relational input data that when transforming will give us V1_StringTestDataSource back. 
This should be for all InputData that has the data input.

File: packages/legend-graph/src/models/protocols/pure/v1/transformation/pureProtocol/serializationHelpers/V1_MappingSerializationHelper.ts [hunk 11]
```diff
@@ -644,30 +647,82 @@ enum V1_InputDataType {
...
+    (val) => V1_deserializeTestDataSource(val),
+  ),
   sourceFlatData: usingModelSchema(
     V1_packageableElementPointerDeserializerSchema,
   ),
 });
 
 const V1_relationalInputData = createModelSchema(V1_RelationalInputData, {
   _type: usingConstantValueSchema(V1_InputDataType.RELATIONAL),
-  data: primitive(),
```
Comment (MauricioUyaguari) [inline packages/legend-graph/src/models/protocols/pure/v1/transformation/pureProtocol/serializationHelpers/V1_MappingSerializationHelper.ts:668]: See other comment about keeping this property in reading it `deserialization`

File: packages/legend-graph/src/models/protocols/pure/v1/engine/execution/V1_ExecutionResult.ts [hunk 12]
```diff
@@ -27,6 +27,18 @@ export class V1_ResultBuilder {
   );
 }
 
+export class V1_MappingTestResult {
+  actual!: object;
+  result!: string;
```
Comment (MauricioUyaguari) [inline packages/legend-graph/src/models/protocols/pure/v1/engine/execution/V1_ExecutionResult.ts:32]: result should be an enum as such
```
public enum TestResult
{
    SUCCESS, FAILURE, ERROR
}
```

File: packages/legend-graph/src/models/protocols/pure/v1/engine/execution/V1_ExecutionResult.ts [hunk 13]
```diff
@@ -27,6 +27,18 @@ export class V1_ResultBuilder {
   );
 }
 
+export class V1_MappingTestResult {
```
Comment (MauricioUyaguari) [inline packages/legend-graph/src/models/protocols/pure/v1/engine/execution/V1_ExecutionResult.ts:30]: I would include the testName, mappingPath, and exception (can leave as object).

File: packages/legend-studio/src/stores/editor-state/element-editor-state/mapping/MappingTestState.ts (hunk 6 above)
Comment (MauricioUyaguari) [reply packages/legend-studio/src/stores/editor-state/element-editor-state/mapping/MappingTestState.ts:635]: We need a change in engine for this? It is already an Enum in engine. should be SUCESS, FAILURE, ERROR

**Comment by MauricioUyaguari** [comment]
Not completely done with the review but 2 major things. The sdlc change is good to go as its not much to review. For the engine change, I have not had a chance to look at but I think either ways based on feedback on this review you will need to make changes to the engine PR. 
1. Let's go ahead and as Pierre would say bite the bullet :) and ensure the mapping test api handles running one, some and all mapping tests. It should take a list of `testIds`. If empty or null we should run all of them. The return type should be a list of `MappingTestResult`. We should add a bit more to the test result as mentioned in the comments. 
2. Let's ensure the InputData model change we made is backward compatible for both engine and studio. I know its a bit annoying but we can't break existing projects :)

And for the generate plan for mapping test, go ahead and kill that button. Hoping that allows us to remove the runtime code.

**Comment by akphi** [comment]
@gs-gunjan @MauricioUyaguari Will close this for now as this implementation will be superseded by the new service test flow that @dave-wathen proposed.
